flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/physical/batch/DynamicPartitionPruningRule.java

```diff
@@ -0,0 +1,920 @@
+
+        List<Integer> acceptedFieldIndices =
+                getAcceptedFieldIndices(factJoinKeys, factCalc, factScan, tableSource);
+
+        return !acceptedFieldIndices.isEmpty();
+    }
+
+    private static List<Integer> getAcceptedFieldIndices(
+            List<Integer> factJoinKeys,
+            @Nullable BatchPhysicalCalc factCalc,
+            BatchPhysicalTableSourceScan factScan,
+            DynamicTableSource tableSource) {
+        List<String> candidateFields;
+        if (factCalc == null) {
+            candidateFields =
+                    factJoinKeys.stream()
+                            .map(i -> factScan.getRowType().getFieldNames().get(i))
+                            .collect(Collectors.toList());
+        } else {
+            // Changing the fact key index in fact table calc output to fact key index in fact
+            // table, and filtering these fields that computing in calc node.
+            RexProgram origProgram = factCalc.getProgram();
+            List<Integer> joinKeysIndexInFactTable = new ArrayList<>();
+            List<RexNode> projectInJoinKeyList =
+                    factJoinKeys.stream()
+                            .map(
+                                    i ->
+                                            origProgram.expandLocalRef(
+                                                    origProgram.getProjectList().get(i)))
+                            .collect(Collectors.toList());
+            List<RexNode> exprList = origProgram.getExprList();
+            for (RexNode rex : projectInJoinKeyList) {
+                if (exprList.contains(rex)
+                        && exprList.get(exprList.indexOf(rex)) instanceof RexInputRef) {
+                    joinKeysIndexInFactTable.add(exprList.indexOf(rex));
+                }
+            }
+
+            if (joinKeysIndexInFactTable.isEmpty()) {
+                return Collections.emptyList();
+            }
+
+            candidateFields =
+                    joinKeysIndexInFactTable.stream()
+                            .map(i -> factScan.getRowType().getFieldNames().get(i))
+                            .collect(Collectors.toList());
+        }
+
+        List<String> acceptedFields =
+                ((SupportsDynamicFiltering) tableSource).applyDynamicFiltering(candidateFields);
+
+        return acceptedFields.stream()
+                .map(f -> factScan.getRowType().getFieldNames().indexOf(f))
+                .collect(Collectors.toList());
+    }
+
+    private static boolean isDimSide(RelNode rel) {
+        DppDimSideFactors dimSideFactors = new DppDimSideFactors();
+        visitDimSide(rel, dimSideFactors);
+        return dimSideFactors.isDimSide();
+    }
+
+    /**
+     * Visit dim side to judge whether dim side has filter condition and whether dim side's source
+     * table scan is non partitioned scan.
+     */
+    private static void visitDimSide(RelNode rel, DppDimSideFactors dimSideFactors) {
```

Comment:
I think the basic logic should be there is a filter on dim table or a filter for the partition field on fact table. If there is a filter on fact table, which join a dim table, the result of this method will also return true.

Reply:
> I think the basic logic should be there is a filter on dim table or a filter for the partition field on fact table. If there is a filter on fact table, which join a dim table, the result of this method will also return true.

I will let it more accurate and efficient in future improvements.